thinc/legacy/loss.py

```diff
@@ -0,0 +1,230 @@
+from ..util import to_categorical, get_array_module
+from typing import Optional, Sequence, Dict, Union, Tuple
+from typing import TypeVar, Generic, Any, cast, List
+from ..types import Floats2d, Ints1d
+
+
+LossT = TypeVar("LossT")
+GradT = TypeVar("GradT")
+GuessT = TypeVar("GuessT")
+TruthT = TypeVar("TruthT")
+IntsOrFloats = Union[Ints1d, Floats2d]
+IntsOrFloatsOrStrs = Union[Ints1d, Floats2d, Sequence[int], Sequence[str]]
+
+
+class Loss(Generic[GuessT, TruthT, GradT, LossT]):  # pragma: no cover
+    """Base class for classes computing the loss / gradient. The class can
+    be initialized with settings if needed. It provides get_loss and
+    get_grad as separate methods to allow calculating them separately. It
+    also provides a __call__ method that returns a tuple of both.
+    """
+
+    def __init__(self, **kwargs: Any) -> None:
+        ...
+
+    def __call__(self, guesses: GuessT, truths: TruthT) -> Tuple[GradT, LossT]:
+        return self.get_grad(guesses, truths), self.get_loss(guesses, truths)
+
+    def get_grad(self, guesses: GuessT, truths: TruthT) -> GradT:
+        ...
+
+    def get_loss(self, guesses: GuessT, truths: TruthT) -> LossT:
+        ...
+
+
+class CategoricalCrossentropy(Loss):
```

Comment:
You import this later on as `LegacyCrossEntropy` to distinguish the legacy class from the normal class in the type annotations. I hadn't really considered the typing when suggesting this legacy proposal. Now I'm thinking, could we make a class `LegacyCrossEntropy` here that inherits from the normal `CategoricalCrossentropy` but overrides the methods? The advantage would be that we wouldn't have to change the typing in the normal `loss.py` file - there we can just call the `LegacyCrossEntropy` constructor and say the returned type is a `CategoricalCrossentropy` so the public interface remains the same.

Also maybe we should call it `LegacyCategoricalCrossEntropy` and `LegacySequenceCrossEntropy` for clarity.

Reply:
You mean to have the `LegacyCategoricalCrossEntropy` in the same `loss.py` as `CategoricalCrossEntropy` ?

Reply:
I guess this design was motivated by avoiding the circular imports. Like `legacy` would need to import from `loss.py` but then `loss.py` would need to import from `legacy/loss.py` to make this work I think. 

Here it would be `from thinc.loss import CategoricalCrossentropy` for the inheritance and then in `loss.py` it would be `from thinc.legacy.loss import LegacyCategoricalCrossentropy`.

Reply:
oh okay right!